# UrsaTech: she256 Mentorship Matching Feature Project
## Project Manager: Elyse Garreau
## Project Developers: Allen Li, Angela Jiang, Jasmine Ang, Justina Chen, Kenny Liao, Taroob Zia
### Spring 2020

This file contains an example usage of our program using data that we created manually intending to mimick randomized mentor and mentee applications. Mentors are named "A_ B_" and mentees are named "C_ D_" for the purpose of organization and clarity. This file will display how functions are used, as well as the functionality of our program.

All of the functions/classes that the code uses are defined in the file "functions.py" and in order to use them here, import all from functions. We will use Pandas to display the raw data prior to cleaning.

## Processing the Data

In [1]:
from functions import *
import pandas as pd

Here is displayed the first few rows from the raw mentor data. The applications are Google forms that must be converted to .csv files. Ensure that the columns of your .csv file are the same and also ordered the same way as the columns below. If you change the application or the order of the columns, you may need to modify functions.py to ensure functionality.

In [2]:
raw_mentordata = pd.read_csv("Data/final/mentor.csv")
raw_menteedata = pd.read_csv("Data/final/mentee.csv")
raw_mentordata.head()

Timestamp First Last  Phone number           Email Address  \
0  5/6/2020 15:30:37    A1   B1    6508899836   egarreau@berkeley.edu   
1  5/6/2020 15:32:31    A2   B2    1234567899         alice@gmail.com   
2  5/6/2020 19:36:49    A3   B3    8058321156  allenli75@berkeley.edu   
3  5/6/2020 22:11:12    A4   B4  234-964-1247    testing123@gmail.com   
4  5/7/2020 10:12:00    A5   B5    5101111111             a@gmail.com   

  When you were an undergraduate student, what major or course of study did you pursue?  \
0                                   Computer Science                                      
1        Electrical Engineering and Computer Science                                      
2                             Electrical Engineering                                      
3                                   Computer Science                                      
4                                   Computer Science                                      

  If you pursued a second major, please indicate which one below.   \
0                                       Data Science                 
1                                           Business                 
2                                   Computer Science                 
3                                  Cognitive Science                 
4                       I did not attend university.                 

  What is your current (or most recent) primary occupation?  \
0                                            Student          
1                                                CTO          
2                                    Product Manager          
3                                 Software Developer          
4                                            Student          

      What are your professional interests/passions?  \
0  Data Science, Machine Learning, Distributed Sy...   
1         Blockchain, Computer Science, Data Science   
2  Computer Science, Data Science, Software Engin...   
3  Blockchain, Computer Science, Data Science, So...   
4                                   Computer Science   

            What are some of your hobbies/interests?  ...  \
0                                 Art, Sports, Music  ...   
1                                     Music, Reading  ...   
2  Art, Sports, Music, Dance, Volunteering, Cooki...  ...   
3  Sports, Music, Volunteering, Outdoor Activitie...  ...   
4                        Sports, Music, Video gaming  ...   

  What is your experience with blockchain? [Community]  \
0             Experience from past projects/academia     
1                                    Very unfamiliar     
2             Experience from past projects/academia     
3             Experience from past projects/academia     
4             Experience from past projects/academia     

  What is your experience with blockchain? [Trading]  \
0                                  Somewhat familiar   
1                                    Very unfamiliar   
2                                  Somewhat familiar   
3             Experience from past projects/academia   
4                   Familiar with amateur experience   

  What is your experience with blockchain? [Legal]  \
0                 Familiar with amateur experience   
1                                  Very unfamiliar   
2                                Somewhat familiar   
3                 Familiar with amateur experience   
4                 Familiar with amateur experience   

   What is your experience with blockchain? [Marketing]  \
0             Experience from past projects/academia      
1                                    Very unfamiliar      
2          Professional or industry-level experience      
3             Experience from past projects/academia      
4             Experience from past projects/academia      

  What is your experience with blockchain? [Entrepreneurship]  \
0                                  Somewhat familiar            
1                        

**output_dataframe** takes in three arguments: the path to your .csv file, a defined list **mentor_columns**, and a default argument which specifies if we are dealing with mentor or mentee data. If you are processing mentor data, the argument is default set to 0. If you are processing mentee data, set the argument to 1. We witness the distinction between these two function calls below.

**mentor_columns** is a list of strings defined in functions.py that contains the simplified column names in order. We simplify column names to 1-2 words to make this more clear. For full definitions of each column, refer to the raw data or the applications.

Lastly, **output_dataframe** cleans the data and makes it easier for use in later functions. For instance, it assigns answers numerical values, such as familiarity of a specific blockchain-related skill.

In [3]:
mentorsdf = output_dataframe("Data/final/mentor.csv", mentor_columns)
mentorsdf.head()

first last         phone                   email  \
0    A1   B1    6508899836   egarreau@berkeley.edu   
1    A2   B2    1234567899         alice@gmail.com   
2    A3   B3    8058321156  allenli75@berkeley.edu   
3    A4   B4  234-964-1247    testing123@gmail.com   
4    A5   B5    5101111111             a@gmail.com   

                                         major                  second_major  \
0                             Computer Science                  Data Science   
1  Electrical Engineering and Computer Science                      Business   
2                       Electrical Engineering              Computer Science   
3                             Computer Science             Cognitive Science   
4                             Computer Science  I did not attend university.   

                  job                                     prof_interests  \
0             Student  Data Science, Machine Learning, Distributed Sy...   
1                 CTO         Blockchain, Computer Science, Data Science   
2     Product Manager  Computer Science, Data Science, Software Engin...   
3  Software Developer  Blockchain, Computer Science, Data Science, So...   
4             Student                                   Computer Science   

                                             hobbies                  meet  \
0                                 Art, Sports, Music  Once every 2-3 weeks   
1                                     Music, Reading           Once a week   
2  Art, Sports, Music, Dance, Volunteering, Cooki...  Once every 2-3 weeks   
3  Sports, Music, Volunteering, Outdoor Activitie...  Once every 2-3 weeks   
4                        Sports, Music, Video gaming  Once every 2-3 weeks   

   ...  community trading  legal marketing  entrepreneurship  \
0  ...          3       1      2         3                 1   
1  ...          0       0      0         0                 0   
2  ...          3       1      1         4                 3   
3  ...          3       3      2         3                 3   
4  ...          3       2      2         3                 3   

                                 skills       goals  \
0                            Experience  A1's Goals   
1                         No preference  A2's Goals   
2  Familiarity or demonstrated interest  A3's Goals   
3                         No preference  A4's Goals   
4                         No preference  A5's Goals   

                       why_blockchain                   blockchain_importance  \
0  Why A1 is interested in blockchain  Why A1 thinks blockchain is important.   
1  Why A2 is interested in blockchain  Why A2 thinks blockchain is important.   
2  Why A3 is interested in blockchain  Why A3 thinks blockchain is important.   
3  Why A4 is interested in blockchain  Why A4 thinks blockchain is important.   
4  Why A5 is interested in blockchain  Why A5 thinks blockchain is important.   

   unlisted_profession  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  

[5 rows x 37 columns]

Notice the third argument is a 1, specifying that we are inputting mentee data. The reason for this third argument is the small differences between the mentor and mentee columns.

In [4]:
menteesdf = output_dataframe("Data/final/mentee.csv", mentee_columns, 1)
menteesdf.head()

first last          phone                  email                   major  \
0    C1   D1   555(333)2344     jules@berkeley.edu  Mechanical Engineering   
1    C2   D2  (222) 444-555  egarreau@berkeley.edu       Cognitive Science   
2    C3   D3  234-5431-4951  egarreau@berkeley.edu     Applied Mathematics   
3    C4   D4     5101111111  egarreau@berkeley.edu        Computer Science   
4    C5   D5     1234567890   ajjiang@berkeley.edu        Computer Science   

                 job                                     prof_interests  \
0         Researcher  Machine Learning, Artificial Intelligence, Pro...   
1        Founder/CEO  Cryptography, Web Development, Mobile Development   
2        Founder/CEO  Blockchain, Data Analytics, Software Engineeri...   
3  Software Engineer                Blockchain, Marketing, Cryptography   
4  Software Engineer  Blockchain, Data Analytics, Software Engineeri...   

                                             hobbies                  meet  \
0                      Reading, Volunteering, Baking           Once a week   
1                              Cooking, Baking, Yoga  Once every 2-3 weeks   
2  Reading, Watching TV, Travel, Politics, Video ...          Once a month   
3                        Sports, Music, Video gaming  Once every 2-3 weeks   
4                Dance, Baking, Travel, Video gaming  Once every 2-3 weeks   

   utc  ... community  trading legal  marketing  entrepreneurship  \
0  9.5  ...         2        3     0          0                 0   
1  9.0  ...         1        1     4          4                 1   
2 -6.0  ...         2        1     0          3                 3   
3  8.0  ...         1        1     0          1                 0   
4 -8.0  ...         2        2     2          2                 2   

                                              skills       goals  \
0  DeFi, Usability, Design, Development, Product,...  C1's Goals   
1           Privacy, Usability, Scalability, Trading  C2's Goals   
2  Bitcoin, Ethereum, Design, Development, Produc...  C3's Goals   
3                        Marketing, Entrepreneurship  C4's Goals   
4  Ethereum, Cryptography, Governance, Usability,...  C5's Goals   

                       why_blockchain                   blockchain_importance  \
0  Why C1 is interested in blockchain  Why C1 thinks blockchain is important.   
1  Why C2 is interested in blockchain  Why C2 thinks blockchain is important.   
2  Why C3 is interested in blockchain  Why C3 thinks blockchain is important.   
3  Why C4 is interested in blockchain  Why C4 thinks blockchain is important.   
4  Why C5 is interested in blockchain  Why C5 thinks blockchain is important.   

   second_major  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  

[5 rows x 36 columns]

Now, the dataframes are more easy to deal with. We now call two new functions, **output_mentors** and **output_mentees** to turn each row of the above dataframes into Mentor and Mentee objects. To better understand these functions, let's try them on a shortened version of menteesdf. 

In [5]:
short_menteesdf = menteesdf.sample(10)
output_mentees(short_menteesdf)

{1: <functions.Mentee at 0x118e30c88>,
 2: <functions.Mentee at 0x118e30d30>,
 3: <functions.Mentee at 0x118e30940>,
 4: <functions.Mentee at 0x118e30b38>,
 5: <functions.Mentee at 0x118e308d0>,
 6: <functions.Mentee at 0x118e303c8>,
 7: <functions.Mentee at 0x118e306d8>,
 8: <functions.Mentee at 0x118e30748>,
 9: <functions.Mentee at 0x118e30be0>,
 10: <functions.Mentee at 0x118e4a080>}

**output_mentees** outputs a dictionary. We must store two dictionaries throughout this process which stores all of the Mentor and Mentee object instances. 

As applicants might have repeat names, we have decided to give each Mentor and Mentee object instance an "ID." This "ID" isn't representative of the applicant's data in any way -- in fact, it is simply an integer given to each applicant. The Mentee instances are assigned positive integers in [1, # Mentees] and the mentor instances are assigned negative integers in [-1, -1* (# Mentors)]. We chose this split between positive and negative integers in order to make the distinction between Mentee and Mentor object instances as clear as possible. 

*Note: We considered changing output_mentees to assign their "ID" to a computed hash of their concatenated data. Typically when giving an object a form of idenfitication, a hash makes the most sense. However, all of the functions (particularly stable marriage) rely on their ID's being positive or negative integers, so if you choose to modify this system of identification, you must also modify stable marriage. Moreover, there is always a chance that two applicants could submit the exact same responses, so this integer system of identification would prevent collisions.*

Notice the difference between the above output of **output_mentees** and the below output of **output_mentors**. The Mentors are assigned negative integers.

In [6]:
short_mentorsdf = mentorsdf.sample(10)
output_mentors(short_mentorsdf)

{-1: <functions.Mentor at 0x118e4a710>,
 -2: <functions.Mentor at 0x118e4a780>,
 -3: <functions.Mentor at 0x118e4a6d8>,
 -4: <functions.Mentor at 0x118e4a208>,
 -5: <functions.Mentor at 0x118e4a668>,
 -6: <functions.Mentor at 0x118e4a550>,
 -7: <functions.Mentor at 0x118e4a240>,
 -8: <functions.Mentor at 0x118e4a7f0>,
 -9: <functions.Mentor at 0x118e4a898>,
 -10: <functions.Mentor at 0x118e4a8d0>}

In [7]:
mentors = output_mentors(mentorsdf)
mentees = output_mentees(menteesdf)

We see that taking mentors[-7] gives us the object instance of this mentor.

In [8]:
mentors[-7]

All  of their form responses are stored as object attributes, and that this object is stored in the dictionary "mentors."

In [9]:
[mentors[-7].first, mentors[-7].email, mentors[-7].crypto]

['A7', 'ajjiang@berkeley.edu', 1]

## Creating Match Scores

Match scores are on a scale of 1-100. Match scores represent numerically how much one mentee fits with a mentor and vice versa. This takes into account all answers to the form aside from the three qualitative application questions: their goals, why they believe blockchain is important, and why they are interested in blockchain. 

If you wish to modify match scores by changing the scale or the weight of each attribute, you may do so without necessarily needing to tweak any other function. We assigned weights based on what we believed seemed most important, after viewing past pairings and pairing feedback.

*Note that match scores are separate for mentors and mentees: that is, if mentee A had a score of 75 for mentor B, mentor B's score for mentee A is not necessarily 75. This is because mentees and mentors can look for different attributes in their potential pairing. Mentee A could specify that they want a mentor who has proficiency in ethereum and governance, but this mentee has no background with these skills. Their most preferred mentor (mentor B) may have a lower score for Mentee A because B marked on their application that they would prefer to mentor someone with familiarity or experience with those skills.*

#### On qualitative application responses... 
We decided that these qualitative questions would not significantly contribute to how well matched the pairings are. For instance, mentors and mentees with different appreciations of the importance of blockchain could still be very well matched. In fact, the difference in their values and goals could possibly benefit the pairing: their differences could help each other gain a new appreciation about this vast and contemporary field. These answers are more relative and it is important to provide applicants with a space to write freely. We have chosen to instead display their answers to these questions in the final output dataframe marking all mentor and mentee pairings. Once matched, the mentors and mentees read each other's responses and a conversation begins.

In [10]:
matchscore(mentees, mentors)

The **matchscore** function takes in two arguments: the **mentees** and **mentors** dictionaries. They must be ordered this way, with mentees dictionary as the first argument and mentors dictionary as the second This function does not return anything. Instead, it modifies each Mentee and Mentor object instance. These instances all have the attribute **scores**, which is a dictionary mapping all possible pairings to match scores.

Take mentee 2. Their **scores** attribute stores a dictionary mapping mentor ID's to match scores. The length of this dictionary is equal to the number of mentors. 

In [11]:
mentees[2].scores

{-1: 28.00019073486328,
 -2: 31.67766952966369,
 -3: 30.544165330774643,
 -4: 43.25019073486328,
 -5: 41.5,
 -6: 13.000762939453125,
 -7: 16.544308381922104,
 -8: 28.295643525965072,
 -9: 14.29488058651195,
 -10: 29.25019073486328,
 -11: 17.213616763844207,
 -12: 42.91930838192211,
 -13: 15.62652587890625,
 -14: 29.83842602898093,
 -15: 29.673828125,
 -16: 16.00019073486328,
 -17: 7.125,
 -18: 31.20970869120796,
 -19: 13.295643525965074,
 -20: 44.00004768371582,
 -21: 28.00152587890625,
 -22: 26.794117647058826,
 -23: 15.000190734863281,
 -24: 24.00152587890625,
 -25: 14.91988058651195,
 -26: 42.29430838192211,
 -27: 41.794117647058826,
 -28: 31.625,
 -29: 26.0244140625,
 -30: 3.0030517578125,
 -31: 3.000762939453125,
 -32: 32.006103515625,
 -33: 19.25,
 -34: 11.25,
 -35: 15.000762939453125,
 -36: 27.54449911678539,
 -37: 30.00019073486328,
 -38: 16.088235294117645,
 -39: 26.25019073486328,
 -40: 38.00152587890625,
 -41: 44.00004768371582,
 -42: 28.00152587890625,
 -43: 26.794117647058

We see the same thing for the mentors, where mentor -2 has a dictionary of positive integer mapping to match score.  

In [12]:
mentors[-2].scores

{1: 50.0,
 2: 20.67766952966369,
 3: 2.1255394796609393,
 4: 12.463834764831844,
 5: 5.625134869915235,
 6: 10.632952411890669,
 7: 4.5625,
 8: 25.919657126719763,
 9: 27.91925251697406,
 10: 25.62607895932188,
 11: 28.29425251697406,
 12: 27.91925251697406,
 13: 25.00026973983047,
 14: 27.00026973983047,
 15: 18.1381067932005,
 16: 20.67766952966369,
 17: 10.338834764831844,
 18: 2.0524271728019903,
 19: 13.001078959321879,
 20: 20.919417382415922,
 21: 27.00026973983047,
 22: 25.000134869915236,
 23: 27.00026973983047,
 24: 18.70970869120796,
 25: 27.625134869915236,
 26: 25.000134869915236,
 27: 1.501078959321879,
 28: 25.000033717478807,
 29: 25.00107895932188,
 30: 5.209708691207961,
 31: 18.276213586400996,
 32: 3.002157918643758,
 33: 3.709708691207961,
 34: 25.00053947966094,
 35: 18.000067434957618,
 36: 1.7951966063807023,
 37: 3.001078959321879,
 38: 25.004315837287514,
 39: 13.838834764831844,
 40: 18.000269739830472,
 41: 1.501078959321879,
 42: 18.55242717280199,
 43: 3.5

## Creating Preference Lists

In order to run the stable marriage algorithm, each applicant must have a preference list. In other words, each applicant object instance must have an attribute **preferences** which is an  list of mentors or mentees ordered from most preferred to least preferred. The order of preference depends on their match scores. Similarly to **matchscore**, **preferencelist** takes in the mentees and mentors dictionaries and only modifies their **preferences** attribute to contain a list of possible matches.

In [16]:
preferencelist(mentees, mentors)

Let's look at the first 10 **preferences** list of mentee 3. We see that they're best matched with mentor -35.

In [18]:
mentees[3].preferences[:10]

[-35, -46, -13, -25, -31, -48, -36, -9, -6, -24]

In [21]:
mentees[3].scores[-35]

72.38235294117646

Mentee 3 has a score of ~72.38 for mentor -35. This is a higher score, which makes sense. What about their least preffered mentor?

In [23]:
mentees[3].preferences[40:]

[-4, -14, -37, -1, -26, -3, -15, -29, -20, -41, -5, -18]

In [24]:
mentees[3].scores[-18]

13.258631674575032

We see that Mentee 3 has a lower score of ~13.26 for Mentor -18, which also makes sense given that -18 is their least preffered mentor. 

## Stable Marriage

At last -- the algorithm that matches mentees and mentors. This algorithm is modified from the original algorithm to take into account lists of mentees and mentors of uneven lenghts. The function creates deep copies of the dictionaries so as not to modify the original ones. This function calls another function **same_numbers** which identifies which dictionary is shorter, and adds "dummy" applicants to this shorter dictionary such that these fake applicants are everyone's least favorite. According to the logic of the stable marriage algorithm, as they are everyone's least favorites, their own preferences should not matter. This function outputs a dictionary mapping mentors to mentees according to ID. 

In [27]:
proposals = stablemarriage(mentees, mentors)
proposals

{-5: 4,
 -7: 12,
 -9: 8,
 -10: 5,
 -11: 21,
 -12: 11,
 -17: 24,
 -18: 7,
 -22: 1,
 -23: 9,
 -25: 34,
 -27: 6,
 -28: 17,
 -31: 36,
 -34: 20,
 -39: 40,
 -40: 19,
 -45: 10,
 -50: 13,
 -51: 26,
 -8: 15,
 -13: 41,
 -21: 49,
 -24: 29,
 -33: 28,
 -36: 14,
 -52: 37,
 -44: 25,
 -46: 23,
 -48: 2,
 -3: 35,
 -49: 22,
 -16: 47,
 -35: 3,
 -42: 39,
 -1: 48,
 -43: 18,
 -47: 32,
 -19: 38,
 -4: 16,
 -32: 33,
 -6: 27,
 -29: 44,
 -14: 42,
 -26: 46,
 -20: 30,
 -38: 45,
 -37: 31,
 -30: 43,
 -2: 50,
 -15: 51,
 -41: 52}

## Ouputting the Final Pairings

The function **output_final** takes in three arguments: the output of stable marriage algorithm and the two cleaned dataframes mentors and mentees. This function retrieves the following data from the dataframes: first names, last names, contact information, and qualitative responses. It aligns each mentor/mentee pairing as a row on a dataframe. We see below that C4D4 is paired with A5B5. This function furthermore identifies the "dummy" applicants created by **stablemarriage** and marks them as unpaired. If an applicant is paired with one of these "dummy" applicants, then this means that there is no stable pairing such that they have a match. This is due to the unequal number of mentors and mentees. This is no way means that these mentors/mentees are anyhow unqualified, but instead it merely shows that all other pairings fit better together. 

One last reason we included qualitative answers in this dataframe is to investigate potential mistakes in pairings. If a mentor and mentee happen to write answers that display a clear reason why they shouldn't be paired, we can see here and make any final changes if needed, maybe referencing their answers in the original raw data. 

In [28]:
output_final(proposals, mentors, mentees)

Mentee First Mentee Last            Mentee Email   Mentee Phone  \
0            C4          D4   egarreau@berkeley.edu     5101111111   
1           C12         D12  allenli75@berkeley.edu     8058321156   
2            C8          D8   egarreau@berkeley.edu    22222222222   
3            C5          D5    ajjiang@berkeley.edu     1234567890   
4           C21         D21   egarreau@berkeley.edu  (444)333-4444   
5           C11         D11  allenli75@berkeley.edu     8058321156   
6           C24         D24   egarreau@berkeley.edu   444-333-2221   
7            C7          D7    ajjiang@berkeley.edu     1234567890   
8            C1          D1      jules@berkeley.edu   555(333)2344   
9            C9          D9  allenli75@berkeley.edu     8058321156   
10          C34         D34   egarreau@berkeley.edu   546-334-3494   
11           C6          D6    ajjiang@berkeley.edu     1234567890   
12          C17         D17          bear@gmail.com       19582819   
13          C36         D36   egarreau@berkeley.edu   554-223-4494   
14          C20         D20   egarreau@berkeley.edu   493-333-4444   
15          C40         D40   egarreau@berkeley.edu   823-583-5820   
16          C19         D19    pparker@berkeley.edu   634-444-4444   
17          C10         D10  allenli75@berkeley.edu     8058321156   
18          C13         D13  allenli75@berkeley.edu     8058321156   
19          C26         D26   egarreau@berkeley.edu   444-444-1111   
20          C15         D15          just@gmail.com     2481927491   
21          C41         D41   egarreau@berkeley.edu       31444444   
22          C49         D49   egarreau@berkeley.edu   650-333-3331   
23          C29         D29   egarreau@berkeley.edu   554-444-4444   
24          C28         D28   egarreau@berkeley.edu   333-333-4444   
25          C14         D14       bobette@gmail.com     9885761512   
26          C37         D37   egarreau@berkeley.edu   394-585-3555   
27          C25         D25   egarreau@berkeley.edu   805-705-1335   
28          C23         D23   egarreau@berkeley.edu   444-444-2222   
29           C2          D2   egarreau@berkeley.edu  (222) 444-555   
30          C35         D35   egarreau@berkeley.edu   452-333-4429   
31          C22         D22   egarreau@berkeley.edu   444-333-3333   
32          C47         D47   egarreau@berkeley.edu   805-705-1335   
33           C3          D3   egarreau@berkeley.edu  234-5431-4951   
34          C39         D39   egarreau@berkeley.edu   352-423-3253   
35          C48         D48   egarreau@berkeley.edu   444-444-1111   
36          C18         D18   egarreau@berkeley.edu      292841827   
37          C32         D32   egarreau@berkeley.edu   222-222-2222   
38          C38         D38   egarreau@berkeley.edu   523-145-4424   
39          C16         D16   egarreau@berkeley.edu     1234819281   
40          C33         D33   egarreau@berkeley.edu   999-333-6666   
41          C27         D27   egarreau@berkeley.edu   650-333-3331   
42          C44         D44   egarreau@berkeley.edu   444-333-3333   
43          C42         D42   egarreau@berkeley.edu   493-333-4444   
44          C46         D46   egarreau@berkeley.edu   444-333-2221   
45          C30         D30   egarreau@berkeley.edu   123-333-3333   
46          C45         D45   egarreau@berkeley.edu   444-444-2222   
47          C31         D31   egarreau@berkeley.edu   444-444-4444   
48          C43         D43   egarreau@berkeley.edu  (444)333-4444   
49     Unpaired    Unpaired                Unpaired       Unpaired   
50     Unpaired    Unpaired                Unpaired       Unpaired   
51     Unpaired    Unpaired                Unpaired       Unpaired   

   (Mentee) What are your main goals for this mentorship program?  \
0                                          C4's Goals               
1                                         C12's Goals               
2                                          C8's Goals               
3               

In order to fully have faith in the functionality of this project, let's investigate a few random matches. 

### C12D12 and A7B7
Mentor A7B7 studied computer science and economics, and they are now a software engineer. Mentee C12D12 is a computer science and data science double major, whose dream job is to become a software engineer. They're both interested in blockhain, crypto, machine learning, web dev, and fintech. They also both like baking. While mentee C12B12 can only meet once every 2-3 weeks, mentor A7B7 wants to meet once a week. They're in the same timezone (PST). They're both introverts, and have similar personalities (ISTJ and ISFJ). The mentor would rather their mentee come in with some familiarity in the skills they practice. Luckily for them, mentee C12D12 has familiarity/amateur experience with all of the skills A7B7 is familiar with. Mentee is interested in some skills including bitcoin, crypto, usability -- and their mentor is at least somewhat familiar with all of these. Overall due to their similar interests, career pursuits, expertise, and personality -- these two would make a great match.

### C26D26 and A51B51
Mentee C26 is a CS and economics major who wants to be a researcher, writer, and COO. Mentor A51 is a Data Science and Economics double major who is a software engineer. They're both interested in business/finance. They have differing hobbies. They both intend to meet once a week, and are apart by 1 hour in terms of time zone. They're both fairly extraverted. A51 prefers that their mentee has some familiarity with the fields they're good at. They're professional/experienced in entrepreneurship, product, development, usability, De-Fi, ethereum, and blockchain. C26 has industry-level experience in bitcoin, ethereum, usability, familiarity in development, product, and entrepreneurship. Therefore, A51's preference for their mentee to have some background knowledge is fulfilled. They want to gain skills in product, governance, and design -- their mentor is professionally experienced in product, and familiar with the other two. Overall their timezones are close, they share a major, and have similar interests/goals. 

### C43D43 and A30B50
Mentee C43 is a Materials Science and Physics double major, who wants to be a professor and a Ph.D student. A30B30 studied EECS and is an investor. These two are a little more different, in terms of personality and interests, but they share similar professional skills. C43 wants to gain experience in investment, trading, legal, and marketing. A30 is familiar with legal and marketing and is an investor. They are in the same time zone (EST and PRT). This match is a little weaker than the two previous ones in terms of interests, but this is likely because C43 has a more unique set of majors/career path and due to their matching time zone. If major and interests matters more, the **matchscores** function can be modified to give these attributes a greater weight. 